### ISSUE TITLE
Download Build Artifacts: : broken file permission mask

### ISSUE THREAD

**Issue by omeshp** [microsoft/azure-pipelines-tasks#6364]
## Environment
- Server - VSTS or TFS on-premises?
    VSTS

## Issue Description
Permissions on the files are lost when downloaded to the agent. We have some files set to be executable.

**Comment by omeshp** [comment]
To support retaining of file permissions we need to do changes across the board, first artifact service needs to support retaining metadata for files, then "publish build artifacts" tasks needs to upload the artifacts with permissions. After which "Download build artifacts" task can be updated to honor permissions.

Previous download artifacts in release was writing all files as executable for some agent configurations which isn't correct thing to do.

@bryanmacfarlane Assigning this issue to you since it first requires retaining of file permissions in build artifact.

**Comment by afeblot** [comment]
We took a more generic approach to work around this:
we just create a tgz of what is supposed to be content of the artifact, and have this tgz in the artifact instead. And expand it on the other side.
But really, we should not have to spend time on such issues...

**Comment by mavinm** [comment]
Any status on this?  It's been over a year since this bug was reported.

**Comment by omeshp** [comment]
@mitchdenny Are we supporting retaining of permissions in pipeline artifacts? I know it's on the agenda but wondering if it's already implemented.

**Comment by mitchdenny** [comment]
At this point we don't do it. I think the first priorities for additional metadata to be stored with files in artifacts are:

1. Execute bit
2. Timestamps (may be important for native build caching scenarios)

Both of these are important for both Pipeline Artifacts and Pipeline Caching. Adding @alexmullans for awareness.

**Comment by keithdoyle9** [comment]
I'm also having this issue. Is there a workaround?

**Comment by mavinm** [comment]
See afeblot@ comment for workaround. Simply make a zip/tgz.

**Comment by keithdoyle9** [comment]
Still not working. Keep getting issues with "File not found" errors even though I list out the files right above the command and it exists.

**Comment by johncrim** [comment]
This is still an issue a year and a half later - seems too significant a problem to have not been addressed, IMO.

**Comment by jmalak** [comment]
The same irritating problem here. Artifacts contains executable and pipeline fails because executables can not be run. It should be fixed as urgent because it is fatal problem. It is main purpose of Artifact use some executable created before.

**Comment by nmiodice** [comment]
We have a quick (and dirty) workaround that involves manually setting the permissions we want:
```
# Setting the scripts to be run as executable
chmod -R $WHATEVER_PERMISSIONS $YOUR_FILE_OR_DIRECTORY
```

Can we get an update on the status?

**Comment by drdamour** [comment]
this is an issue with pipeline artifacts too.  AND it causes issues with terraform because best practice of terraform is to artifact and restore the whole .terraform directory and the x bits on the providers are lost

**Comment by jahead** [comment]
@mitchdenny Are we able to get an update on this? It's been over a year, and no feedback?

**Comment by mitchdenny** [comment]
Adding @johnterickson

**Comment by ngarg-panw** [comment]
I had installed a version of my extension 6 months back and now I again installed the updated version. 
In the release pipeline for 6 months back extension, i had used the sudo command and chmod command which seems to work as expected. But the same commands doesn't work in the command line task now.
Old one:
command line:
`sudo mv $(System.DefaultWorkingDirectory)/twistcli /usr/bin/twistcli
chmod +x /usr/bin/twistcli
ls -la /usr/bin/twistcli`

Output:
`2020-05-26T23:59:12.1911901Z ##[section]Starting: Move twistcli to binary path and make executable
2020-05-26T23:59:12.1985635Z ==============================================================================
2020-05-26T23:59:12.2006779Z Task         : Command line
2020-05-26T23:59:12.2015264Z Description  : Run a command line script using Bash on Linux and macOS and cmd.exe on Windows
2020-05-26T23:59:12.2023654Z Version      : 2.164.1
2020-05-26T23:59:12.2028755Z Author       : Microsoft Corporation
2020-05-26T23:59:12.2036370Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/command-line
2020-05-26T23:59:12.2059796Z ==============================================================================
2020-05-26T23:59:12.3944815Z Generating script.
2020-05-26T23:59:12.3968967Z ========================== Starting Command Output ===========================
2020-05-26T23:59:12.4016397Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/78e4fd0e-2a1e-44cb-bb6b-b3c084a21dc9.sh
2020-05-26T23:59:12.4337689Z -rwxr-xr-x 1 vsts docker 32132568 May 26 23:59 /usr/bin/twistcli`

Now:
CMD Script:
`mv $(System.DefaultWorkingDirectory)/twistcli /usr/bin/twistcli
chmod 777 /usr/bin/twistcli
ls -la /usr/bin/twistcli
docker pull nginx:latest`


output:
`2020-05-27T00:05:19.0704491Z -rw-r--r-- 1 VssAdministrator 197121 32373360 May 27 00:04 /usr/bin/twistcli
2020-05-27T00:05:19.1386671Z -rw-r--r-- 1 VssAdministrator 197121 32373360 May 27 00:04 /usr/bin/twistcli
2020-05-27T00:05:25.7218133Z latest: Pulling from library/nginx
2020-05-27T00:05:25.7218891Z no matching manifest for windows/amd64 10.0.14393 in the manifest list entries
2020-05-27T00:05:25.8131849Z ##[error]Cmd.exe exited with code '1'.`

If you compare from older output, the chmod has not worked.
Also, the user looks different.

Pls note: You don't see sudo here as it doesn't work. The error is something like unknow command sudo.

I have spent whole day figuring out what's wrong with no luck.

**Comment by drdamour** [comment]
Stale but important

**Comment by jmalak** [comment]
It was already resolved by adding note in artifact documentation that permission are not preserved due to Windows filesystem.
Don't expect any solution. It works as designed. Nice way how to fix problem 🥇

**Comment by jahead** [comment]
@jmalak Maybe I've missed something in the history of this issue and the issues referencing this one. But no where was it communicated as resolved AFAIK or working as designed. 

We have a workaround

> @mitchdenny Are we supporting retaining of permissions in pipeline artifacts? I know it's on the agenda but wondering if it's already implemented.

Would suggest that there was the intent of at least the possibility supporting this natively.

Hopefully we get some feedback soon

**Comment by jmalak** [comment]
It should be resolved proper way, what I commented is Microsoft quick solution.
I am using my own solution that I encapsulate artifact into archive with permission and if I need it I restore it from archive (7z format on Windows and tar.gz on NIX).

**Comment by tknerr** [comment]
Oh damn, also ran into this today :-/

Our use case is build parallelization: so we are currently splitting our build and running it as several pipeline jobs in parallel, publishing the output directories as pipeline artifacts. If all parallel jobs complete, we are collecting the output directories from each job (i.e. download pipeline artifacts) and merging them back into one directory. This is where all the file permissions are getting lost :-/

I could imagine this is quite a common use case / scenario, so I would really appreciate if this would be supported out-of-the-box with Azure DevOps pipeline artifacts.

**Comment by jmalak** [comment]
No interest from MS guys, read date of starting this issue.

**Comment by ChristianUlbrich** [comment]
**FWIW** _cross-platform_ artifact upload is possible with the the hosted build agents, because they all have `tar` on board. Thus we _"simply"_ _tar_ all artifacts with `tar cvpzf ../browser-cache.tgz .`. This will not only retain permissions, but works with _symbolic_ links as well (which are not supported with `PublishPipelineArtifact@1` either).

For restoring, we use:
```yaml
- bash: |
      tar xvf browser-cache.tgz
    workingDirectory: '$(Agent.BuildDirectory)/browser-cache'
    displayName: 'Extract browser cache'
```

There are various other bugs / nuisances around artifacts; `$(Agent.BuildDirectory)` is interpreted differently on containers (and so on), but with enough patience you can cobble something together.

But I totally agree, that "tarring" should be an optional option for uploading artifacts to make them worth the effort for cross-platform builds.

**Comment by DaniilShmelev** [comment]
Hey everyone! We've added optional inputs to PublishBuildArtifacts and DownloadBuildArtifacts tasks to store artifacts as tar archives, which preserve file permissions.
These changes are going to be deployed over the next few weeks.

**Comment by afeblot** [comment]
That's very cool, thanks.
That said, those tasks are now deprecated in favor of PublishPipelineArtifacts and DownloadPipelineArtifacts. Are you going to apply the same change to these new tasks? Or were they not affected by the issue?

**Comment by drdamour** [comment]
PublishPipelineArtifacts  has the same issue

**Comment by nzcodarnoc** [comment]
It's 2021 and I've just found this issue because I've run into the same problem. I guess I'm going to tar everything before the publish step.

**Comment by blacksails** [comment]
Can we reopen this please and get it fixed for the `publish` shorthand and `PublishPipelineArtifacts`?

**Comment by drdamour** [comment]
@blacksails they almost never reopen, better to open a new ticket and mention this one

**Comment by KUGA2** [comment]
Has anyome bothered to open a new issue? I guess, until then, the workaround is to use the deprecated tasks ([PublishBuildArtifacts@1](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/publish-build-artifacts-v1?view=azure-pipelines), [DownloadBuildArtifacts@1 ](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/download-build-artifacts-v1?view=azure-pipelines)) as they do support:

```
    #StoreAsTar: false # boolean. Tar the artifact before uploading. Default: false.
```
or:
```
    #extractTars: # boolean. Extract all files that are stored inside tar archives.
```

They are still here and do still work, right?
